rust/azure_iot_operations_services/src/leased_lock/lease.rs

```diff
@@ -0,0 +1,307 @@
+    ///
+    /// [`struct@Error`] of kind [`MissingFencingToken`](ErrorKind::MissingFencingToken) if the fencing token in the service response is empty.
+    ///
+    /// # Panics
+    /// Possible panic if, for some error, the fencing token (a.k.a. `version`) of the acquired lease is None.
+    pub async fn acquire(
+        &mut self,
+        lease_expiration: Duration,
+        request_timeout: Duration,
+        renewal_period: Option<Duration>,
+    ) -> Result<HybridLogicalClock, Error> {
+        // Stop auto-renewal.
+        self.auto_renewal_cancellation_token.cancel();
+
+        let acquire_result = self
+            .internal_acquire(lease_expiration, request_timeout)
+            .await;
+
+        if let Some(renewal_period) = renewal_period {
+            self.auto_renewal_cancellation_token = CancellationToken::new();
+            let self_clone = self.clone();
+
+            tokio::task::spawn({
+                async move {
+                    loop {
+                        select! {
+                            () = self_clone.auto_renewal_cancellation_token.cancelled() => {
+                                break; // Auto-renewal is cancelled.
+                            }
+                            () = tokio::time::sleep(renewal_period) => {
+                                if self_clone
+                                    .internal_acquire(lease_expiration, request_timeout)
+                                    .await
+                                    .is_err()
+                                {
+                                    // Acquire failed. Stopping Auto-renewal.
+                                    break;
+                                }
+                            }
+                        }
+                    }
+                }
+            });
+        }
+
+        acquire_result
+    }
+
+    /// Releases a lease if and only if requested by the lease holder (same client id).
+    ///
+    /// Note: `request_timeout` is rounded up to the nearest second.
+    ///
+    /// Returns `Ok()` if lease is no longer held by this `lease_holder`, or `Error` otherwise.
+    /// # Errors
+    /// [`struct@Error`] of kind [`InvalidArgument`](ErrorKind::InvalidArgument) if the `request_timeout` is zero or > `u32::max`
+    ///
+    /// [`struct@Error`] of kind [`ServiceError`](ErrorKind::ServiceError) if the State Store returns an Error response
+    ///
+    /// [`struct@Error`] of kind [`UnexpectedPayload`](ErrorKind::UnexpectedPayload) if the State Store returns a response that isn't valid for a `V Delete` request
+    ///
+    /// [`struct@Error`] of kind [`AIOProtocolError`](ErrorKind::AIOProtocolError) if there are any underlying errors from the command invoker
+    ///
+    /// # Panics
+    /// Possible panic if `std::Lock::lock()` fails.
+    pub async fn release(&mut self, request_timeout: Duration) -> Result<(), Error> {
```

Comment:
don't think you'll need &mut self here either! (and then that will extend to unlock() as well)